### ISSUE TITLE
Reconsider use of jquery().data() methd

### ISSUE THREAD

**Issue by karger** [zepheira/exhibit3#99]
With the move to data- attributes there is some use of the jquery().data() method in exhibit.  This is inferior to the use of jquery.attr('data-...') for two reasons.
1.  .data() attempts to _evaluate_ the value the attribute (using parseJSON).  Thus, a function expecting a string might receive a boolean or an object.  .attr is guaranteed to return a string
2.  .attribute tries to "normalize" attribute names, for example converting camelcase to hyphenated words (viewClass ---> view-class) which creates some robustness

As a particular example of the way this introduced a bug into the current exhibit, Exhibit.viewPanel.createFromDOM looks up the "viewClass" attribute.  Exhibit.getAttribute translates this into .data('ex-viewClass") which will not properly match a data-ex-view-class attribute OR a data-ex-viewclass attribute.

Thus, I recommend reverting all uses of jquery.data to use jquery.attr instead.

**Comment by zepheiryan** [comment]
1. Cases where this is an issue should be resolved instead of throwing the whole thing out.  I'd rather expect to store and retrieve a number than store a number and parse it later out of a string.  Writing a function expecting a string should have a matching function storing a string.  It rather boils down to the same thing, but an intelligent storage mechanism sounds more useful.
2. So [does data](http://api.jquery.com/data/#data-html5).

**Comment by karger** [comment]
Using .data as a storage mechanism for computational state on a node is very sensible.  But we never use that capability in exhibit.  What I am questioning is its use for retrieving configuration attributes of dom nodes, which we know are strings (so your first point doesn't apply).  The "caching" behavior of the .data() method means this makes exhibit insensitive to changes in the html.  Since one of the main contributions of exhibit is the way it describes all configuration in the html document itself, it seems undesirable to interfere with that behavior.

**Comment by karger** [comment]
Note also that according to the html5 spec, updates to a nodes "dataset" object (the dom object that is populated by .data-\* attributes) must be reflected in changes to the corresponding attributes on the dom node (see http://www.w3.org/html/wg/drafts/html/master/dom.html#embedding-custom-non-visible-data-with-the-data-*-attributes ), section 3.2.3.9, step 6 of the algorithm for setting names to certain values.  jquery doesn't promise to follow the html5 spec, but this is a case where their implementation is pretty clearly in conflict with the spec's intent.

**Comment by karger** [comment]
I've looked at this some more and had some discussion on the jquery mailing list, and am convinced that we should switch to .attr('data-foo') for accessing data elements.  jquery's .data() method conflates two issues: reading data-foo attributes, which we use a lot in exhibit, and storing/retrieving node-specific information, which we do not use at all.  For the latter task, the .data() method is exactly right.  Transforming things to/from strings for storage as attributes is inefficient (and impossible for some data types) so should be avoided.  However, the ONLY way we use the .data() method in exhibit is to read attributes; we never use it to store or access stored data.

Using .data() provides no gains in efficiency, because we only read each attribute once.  It doesn't benefit from .data()s type-coercion because exhibit already explicitly handles such coercion through its settingSpecs module.  It also creates a potential for future bugs of the sort I opened this issue with: by accessing a cached version of the attribute, it can fail to get an up to date value.  I see no benefit to demanding that all developers over exhibit handle this when we can ensure, by using .attr('data-foo'), that the data exhibit uses is up to date.

So, in the absence of any objections, I'm going to put together a pull request that switches out .data() for .attr('data-foo').  Note that this does not preclude using .data() in the future to store and retrieve node-specific information; it just establishes that as a separate functionality.
